mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseToStdlib.cpp

```diff
@@ -1456,45 +1467,122 @@ struct XdlopsGemmV2RewritePattern : public OpRewritePattern<XdlopsGemmV2Op> {
+        }
       } else {
-        argA = innerLoopb.create<memref::LoadOp>(loc, argType, op.bufferA(),
-                                                 ValueRange{offset});
-        argB = innerLoopb.create<memref::LoadOp>(loc, argType, op.bufferB(),
-                                                 ValueRange{offset});
+        auto constructArg = [&outerLoopb, &innerLoopb, &loc,
+                             &b](Value &buffer, Value &arg, Value &outerLoopiv,
+                                 Value &innerLoopiv, Type &argType,
+                                 int64_t k_base) {
+          arg = createZeroConstantOp(innerLoopb, loc, argType);
+
+          Value argAWide = innerLoopb.create<memref::LoadOp>(
+              loc,
+              buffer.getType()
+                  .cast<MemRefType>()
+                  .getElementType()
+                  .cast<VectorType>(),
+              buffer, ValueRange{outerLoopiv});
+
+          for (int64_t i = 0; i < k_base; ++i) {
+            Value iterOp = innerLoopb.create<ConstantIndexOp>(loc, i);
+            Value newOffset =
+                innerLoopb.create<AddIOp>(loc, innerLoopiv, iterOp);
+            auto element = innerLoopb.create<vector::ExtractElementOp>(
+                loc, argType.cast<VectorType>().getElementType(), argAWide,
+                newOffset);
+            arg = innerLoopb.create<vector::InsertElementOp>(loc, element, arg,
+                                                             iterOp);
+          }
+        };
+
+        auto bufferAVectorLen =
+            bufferAElementType.cast<VectorType>().getShape()[0];
+        auto bufferBVectorLen =
+            bufferBElementType.cast<VectorType>().getShape()[0];
+        if ((bufferAVectorLen < k_base) || (bufferAVectorLen % k_base != 0)) {
+          llvm_unreachable(
+              "bufferA Vector length must be divisible by xdlops argument "
+              "length");
+        }
+        if ((bufferBVectorLen < k_base) || (bufferBVectorLen % k_base != 0)) {
+          llvm_unreachable(
+              "bufferB Vector length must be divisible by xdlops argument "
+              "length");
+        }
+
+        if (k_base == bufferAVectorLen && k_base == bufferBVectorLen) {
+          // If xdlops argument vector size happen to be similar to the vgpr
+          // vector length. Use memref load directly
+          argA = innerLoopb.create<memref::LoadOp>(
+              loc, bufferAElementType.cast<VectorType>(), bufferA,
+              ValueRange{outerLoopiv});
+          argB = innerLoopb.create<memref::LoadOp>(
+              loc, bufferBElementType.cast<VectorType>(), bufferB,
+              ValueRange{outerLoopiv});
+        } else {
+          // If xdlops argument vector size is smaller than the vgpr vector
+          // length, we have to temporarily construct smaller vgpr vector
+          // for xdlops to consume
+          constructArg(bufferA, argA, outerLoopiv, innerLoopiv, argType,
```

Comment:
how about change to:

`argA = constructArg(bufferA, outerLoopiv, innerLoopiv, argType, k_base)` to make it more comprehensible about the input / output?

Reply:
Sure